### DISCUSSION TITLE
OPFS without nativeFS

### DISCUSSION THREAD

**Discussion by airen1986** [pyodide/pyodide#5109] [Q&A]
Hi Team,

Is it possible to access OPFS file sysem without nativeFS API, as it is not supported in Firefox?

Thanks

**Comment by ryanking13** [comment]
> Is it possible to access OPFS file sysem without nativeFS API, as it is not supported in Firefox?

Yes and no

You need to use nativeFS API to use OPFS file system, but the parameter that the nativeFS API accepts is [FileSystemHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle). So you can use OPFS apis such as [StorageManager](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory) to get the handle even in Firefox.

**Comment by airen1986** [comment]
I tried using storage handle in Firefox as shown in code below, but it didn't work. Same code worked in Chrome

```
async function loadPyodideAndPackages() {
  let root = await navigator.storage.getDirectory();
  let dirHandleMount = await root.getDirectoryHandle('data', { create: true });

  self.pyodide = await loadPyodide({
    indexURL:"https://cdn.jsdelivr.net/pyodide/v0.26.2/full",
  });

  await self.pyodide.loadPackage(["micropip", "numpy", "pandas", "pytz", "tzdata", "six", "python-dateutil"]);
  const micropip = self.pyodide.pyimport("micropip");
  await self.pyodide.mountNativeFS("/mnt/nativefs", dirHandleMount);
  await self.pyodide.runPythonAsync(package_code);
};

```

**Comment by jburgy** [comment]
I saw this discussion _after_ asking a [similar question](https://discord.com/channels/1254759875347611680/1428349062653939845) on discord.  Difference is I was trying to do it in python instead of JavaScript.  That's when I noticed that `pyodide` exposes its js module to python as [`pyodide_js`](https://pyodide.org/en/stable/usage/api/js-api.html#module-pyodide) so I thought I could simply write
```python
handle = await js.navigator.storage.getDirectory()
await pyodide_js.mountNativeFS("/opfs", handle)
``` 

Unfortunately (because there's always a "but"), when I tried
```python
with open("/opfs/test.txt", "wt") as file:
    print("Hello, World!", file=file)
```
I got no python error but a few
```
registry.js:317 Could not resolve location of /opfs on server
resolvePath @ registry.js:317
renderers.js:915 Path resolution bailing: does not exist
```
in the DevTools console and [OPFS Explorer](https://chromewebstore.google.com/detail/opfs-explorer/acndjpgkpaclldomagafnognkcgjignd?hl=en) confirms that my OPFS remained empty.  What should I do?

**Reply by jburgy** [reply]
Ignore me, that was an [ID:10T error](https://en.wikipedia.org/wiki/User_error#ID-10-T_error) because I forgot to `await nativefs.syncfs()` after the fact!
